dev-docs/tls.md

```diff
@@ -0,0 +1,152 @@
+
+Create a PKCS12 file. PKCS12 files are primarily useful for creating instances
+of a Java `KeyStore`. It is not possible to [elide a
+password](https://stackoverflow.com/questions/27497723/export-a-pkcs12-file-without-an-export-password)
+empty passwords are not recommended.
+
+```
+openssl pkcs12 -export \
+        -out server-keystore.p12 \
+        -inkey server-key.pem \
+        -in server-cert.pem \
+        -passout pass:foobar
+```
+
+## Inspect a Certificate
+
+Print the start and end dates for a given certificate. Note that a certificate
+whose start date is in the future is called "expired" by many tools.
+
+```
+openssl x509 -startdate -enddate -noout -in cert.pem
+```
+
+Print a complete textual representation of the certificate.
+
+```
+openssl -text -noout -in cert.pem
+```
+
+## Determine the Cause of Certificate Expiration
+
+Check the start and end dates.
+
+```
+openssl x509 -startdate -enddate -noout -in cert.pem
+```
+
+Is the start date in the future? Is the end date in the past? What is the time
+on the machine on which the certificate failure occurred? Remember to always
+compare times in UTC!
+
+All Hail certificates are signed by a root certificate named `hail-root`. If
+this certificate is expired, the signed certificate will also be called
+"expired" even though it is not itself expired. Download the Hail root
+certificate to a local file and inspect the start and end dates.
+
+```
+kubectl get secrets ssl-config-hail-root -o json \
+    | jq -r '.data["hail-root-cert.pem"]' \
+    | base64 -D \
+    > hail-root-cert.pem
+```
+
+This certificate should be bit-for-bit identical to the `SERVICE-incoming.pem`
+trust file, but you should verify that. For example, download the
+`router` incoming trust:
+
+```
+kubectl get secrets ssl-config-router -o json \
+    | jq -r '.data["router-incoming.pem"]' \
+    | base64 -D \
+    > hail-root-cert.pem
+```
+
+## Regenerate All the Certificates
+
+If something has gone wrong, a relatively straightforward way to get back to
+working is to regenerate all the certificates. This procedure will cause
+downtime: services using the old certs will not trust servers using the new
+certs and vice-versa. Once all services have restarted, there should be no
+downtime.
+
+1. Regenerate the root certificate (from your laptop):
+
+```
+openssl req -new -x509 \
+        -subj /CN=hail-root \
+        -nodes \
+        -newkey rsa:4096 \
+        -keyout hail-root-key.pem \
+        -out hail-root-cert.pem \
+        -days 365
+```
+
+2. Update kubernetes with the new root certificate:
+
+```
+kubectl create secret generic \
+        -n default ssl-config-hail-root \
+        --from-file=hail-root-key.pem \
+        --from-file=hail-root-cert.pem \
+        --dry-run -o yaml \
+    | kubectl apply -f -
+```
+
+3. Update all the service certificates:
+
+```
+python3 /PATH/TO/HAIL/tls/create_certs.py \
+        default \
+        /PATH/TO/HAIL/tls/config.yaml \
```

Comment:
I like $HAIL_HOME.  If we standardize on that and people set it, then this would work copy/paste.